core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,306 @@
+import io.netty.handler.codec.dns.DnsRecord;
+import io.netty.handler.codec.dns.DnsRecordType;
+import io.netty.resolver.AbstractAddressResolver;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+
+final class RefreshingAddressResolver extends AbstractAddressResolver<InetSocketAddress> {
+
+    private static final Logger logger = LoggerFactory.getLogger(RefreshingAddressResolver.class);
+
+    private final ConcurrentMap<String, CompletableFuture<CacheEntry>> cache;
+    private final DefaultDnsNameResolver resolver;
+    private final List<DnsRecordType> dnsRecordTypes;
+    private final int minTtl;
+    private final int maxTtl;
+    private final Backoff refreshBackoff;
+
+    private volatile boolean resolverClosed;
+
+    RefreshingAddressResolver(EventLoop eventLoop,
+                              ConcurrentMap<String, CompletableFuture<CacheEntry>> cache,
+                              DefaultDnsNameResolver resolver, List<DnsRecordType> dnsRecordTypes,
+                              int minTtl, int maxTtl, Backoff refreshBackoff) {
+        super(eventLoop);
+        this.cache = cache;
+        this.resolver = resolver;
+        this.dnsRecordTypes = dnsRecordTypes;
+        this.minTtl = minTtl;
+        this.maxTtl = maxTtl;
+        this.refreshBackoff = refreshBackoff;
+    }
+
+    @Override
+    protected boolean doIsResolved(InetSocketAddress address) {
+        return !requireNonNull(address, "address").isUnresolved();
+    }
+
+    @Override
+    protected void doResolve(InetSocketAddress unresolvedAddress, Promise<InetSocketAddress> promise)
+            throws Exception {
+        requireNonNull(unresolvedAddress, "unresolvedAddress");
+        requireNonNull(promise, "promise");
+        if (resolverClosed) {
+            promise.tryFailure(new IllegalStateException("resolver is closed already."));
+            return;
+        }
+        final String hostname = unresolvedAddress.getHostString();
+        final int port = unresolvedAddress.getPort();
+        final CompletableFuture<CacheEntry> entryFuture = cache.get(hostname);
+        if (entryFuture != null) {
+            handleFromCache(entryFuture, promise, port);
+            return;
+        }
+
+        final CompletableFuture<CacheEntry> result = new CompletableFuture<>();
+        final CompletableFuture<CacheEntry> previous = cache.putIfAbsent(hostname, result);
+        if (previous != null) {
+            handleFromCache(previous, promise, port);
+            return;
+        }
+
+        final List<DnsQuestion> questions =
+                dnsRecordTypes.stream()
+                              .map(type -> DnsQuestionWithoutTrailingDot.of(hostname, type))
+                              .collect(toImmutableList());
+        sendQueries(questions, hostname, result);
+        result.handle((entry, cause) -> {
+            if (cause != null) {
+                cache.remove(hostname);
```

Comment:
I wonder removing the cached result is a good strategy. IMHO, I want to get cached DNS if there is an error on the DNS result.

Reply:
Then, the client will never get resolved addresses.

Reply:
Add `negativeTtl` to give a choice to users to cache or not when DNS queries are failed. The default value is `0` which means the resolve does not cache when the query fails, so it will be no problem. 😄

Reply:
@anuraaga and @trustin PTAL again. 😉